src/test/spec/unified_runner/operation.rs

```diff
@@ -2613,6 +2643,100 @@ impl TestOperation for CreateEntities {
     }
 }
 
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub(super) struct Download {
+    id: Bson,
+}
+
+impl TestOperation for Download {
+    fn execute_entity_operation<'a>(
+        &'a self,
+        id: &'a str,
+        test_runner: &'a TestRunner,
+    ) -> BoxFuture<'a, Result<Option<Entity>>> {
+        async move {
+            let bucket = test_runner.get_bucket(id).await;
+            let mut stream = bucket.open_download_stream(self.id.clone()).await?;
+            let mut buf = Vec::new();
+            stream.read(&mut buf).await?;
+            Ok(Some(Entity::Bson(Bson::String(
+                String::from_utf8(buf).unwrap().into(),
+            ))))
+        }
+        .boxed()
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub(super) struct DownloadByName {
+    filename: String,
+    #[serde(flatten)]
+    options: GridFsDownloadByNameOptions,
+}
+
+impl TestOperation for DownloadByName {
+    fn execute_entity_operation<'a>(
+        &'a self,
+        id: &'a str,
+        test_runner: &'a TestRunner,
+    ) -> BoxFuture<'a, Result<Option<Entity>>> {
+        async move {
+            let bucket = test_runner.get_bucket(id).await;
+            let mut stream = bucket
+                .open_download_stream_by_name(self.filename.clone(), self.options.clone())
+                .await?;
+            let mut buf = Vec::new();
+            stream.read(&mut buf).await?;
+            Ok(Some(Bson::String(String::from_utf8(buf).unwrap()).into()))
+        }
+        .boxed()
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub(super) struct Upload {
+    source: Document,
+    filename: String,
+    // content_type is deprecated and no longer supported.
+    // Option included for deserialization.
+    #[serde(rename = "contentType")]
+    _content_type: Option<String>,
+    #[serde(rename = "disableMD5")]
+    _disable_md5: Option<bool>,
+    #[serde(flatten)]
+    options: GridFsUploadOptions,
+}
+
+impl TestOperation for Upload {
+    fn execute_entity_operation<'a>(
+        &'a self,
+        id: &'a str,
+        test_runner: &'a TestRunner,
+    ) -> BoxFuture<'a, Result<Option<Entity>>> {
+        async move {
+            let bucket = test_runner.get_bucket(id).await;
+            let hex_bytes = self.source.get("hexBytes").unwrap().as_str().unwrap();
+            let hex_string: Vec<u8> = (0..hex_bytes.len())
+                .step_by(2)
+                .map(|i| u8::from_str_radix(&hex_bytes[i..i + 2], 16).unwrap())
+                .collect();
```

Comment:
This is creating a range with a 2 step increment (e.g. 0, 2, 4, 6,...) and then taking pairs of characters from `$$hexBytes` and mapping them to a single byte, since two characters in a hex string would be a byte. Finally it collects them into a `Vec<u8>`.